File: cache/sum-matrix/sum-matrix.rb
N = 4096

def sum_by_row(m)
  t1 = Time.now
  s = 0.0
  for row in 0 ... N do
    for col in 0 ... N do
      s += m[row*N + col]
    end
  end
  t2 = Time.now
  puts "sum_by_row (#{s}): #{1e6*(t2-t1)} micro-seconds"
end

def sum_by_col(m)
  t1 = Time.now
  s = 0.0
  for row in 0 ... N do
    for col in 0 ... N do
      s += m[row + col*N]
    end
  end
  t2 = Time.now
  puts "sum_by_col (#{s}): #{1e6*(t2-t1)} micro-seconds"
end

t1 = Time.now
m = [0] * (N*N)
for i in 0 ... N*N do
  m[i] = i % 2
end
t2 = Time.now
puts "init_matrix: #{1e6*(t2-t1)} micro-seconds"

for c in ARGV[0].each_char do
  case c
  when 'r'
    sum_by_row(m)
  when 'c'
    sum_by_col(m)
  end
end

File: cache/sum-matrix/sum-matrix.go
package main

import (
	"fmt"
	"os"
	"time"
)

const N int = 4096

func initMatrix() []float64 {
	t1 := time.Now()

	m := make([]float64, N*N)
	for i := 0; i < len(m); i++ {
		m[i] = float64(i % 2)
	}

	dur := time.Since(t1)
	fmt.Printf("initMatrix: %d micro-seconds\n", dur.Nanoseconds() / 1000)

	return m
}

func sumByRow(m []float64) {
	t1 := time.Now()

	s := 0.0;
	for row := 0; row < N; row++ {
		for col := 0; col < N; col++ {
			s += m[row*N + col]
		}
	}

	dur := time.Since(t1)
	fmt.Printf("sumByRow (%f): %d micro-seconds\n", s, dur.Nanoseconds() / 1000)
}

func sumByCol(m []float64) {
	t1 := time.Now()

	s := 0.0;
	for row := 0; row < N; row++ {
		for col := 0; col < N; col++ {
			s += m[row + col*N]
		}
	}

	dur := time.Since(t1)
	fmt.Printf("sumByCol (%f): %d micro-seconds\n", s, dur.Nanoseconds() / 1000)
}

func main() {
	m := initMatrix()

	for _, s := range os.Args[1:] {
		for _, c := range s {
			switch c {
			case 'r': sumByRow(m)
			case 'c': sumByCol(m)
			}
		}
	}
}
